### PR TITLE
add ability to disable ability to override values of existing fields in set processor

### LINKED ISSUE #17659: Default value for a field
**Describe the feature**:
Currently, we have null_value, which works fine as long as the value is declared to be null at index time.  The `set` processor overwrites values if they already exist.  It'd be nice to have the ability to have the ability to set a value in the missing case.  To me, it seems like it'd be nice to add to the ingest processors (perhaps a simple `no_override` field in `set`?) so we can have different defaults depending on the pipeline

Related to https://github.com/elastic/elasticsearch/issues/12998.  Has been requested on discuss as well a few times:
https://discuss.elastic.co/t/is-it-possible-to-index-a-default-value-when-a-document-property-is-undeclared/9849
https://discuss.elastic.co/t/setting-missing-field-to-default-value-based-on-a-master-template/39317
https://discuss.elastic.co/t/setting-a-default-value-for-missing-fields-during-indexing/45021
https://discuss.elastic.co/t/field-null-value-doesnt-seem-to-have-any-effect/22015

### PR DESCRIPTION
**Pull request by talevy**
Fixes #17659

adds `override` option for the `set` processor. When set to `false`, the processor will not update 
the value of an existing non-null-valued field in the ingest document

```
{
  "set" : {
    "field" : ...,
    "value" : ...,
    "override" : (true|false)
  }
}
```

This field is optional, and defaults to `true`.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/ingest/processor/SetProcessor.java [hunk 1]
```diff
@@ -55,7 +65,9 @@ public ValueSource getValue() {
 
     @Override
     public void execute(IngestDocument document) {
-        document.setFieldValue(field, value);
+        if (overrideEnabled || !document.hasField(field) || document.getFieldValue(field, Object.class) == null) {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/processor/SetProcessor.java]: Can you change `!document.hasField(field)` into `document.hasField(field) == false`. This is kind of the standard now in the code base, because it is easy to overlook the `!`.

File: core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java [hunk 2]
```diff
@@ -142,6 +154,16 @@ public IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, String>
     }
 
     /**
+     * Checks whether the document contains a value for the provided templated path
+     * @param fieldPathTemplate the template for the path within the document in dot-notation
+     * @return true if the document contains a value for the field, false otherwise
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java]: I think we also return `true` if the field exists but there is no value? (`null` value) Maybe we should mention this too?

**Comment by martijnvg** [comment]
Left 2 small remarks, LGTM otherwise.

File: core/src/main/java/org/elasticsearch/ingest/processor/SetProcessor.java (hunk 1 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/processor/SetProcessor.java]: yes

File: core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java (hunk 2 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java]: maybe this is outside the scope of this PR? as this `hasField` simply overrides the first to enable the `Template` argument

Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/core/IngestDocument.java]: true
